servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcClientBuilder.java

```diff
@@ -168,4 +227,91 @@ public void doAppendHttpClientFilter(final Predicate<StreamingHttpRequest> predi
                                          final StreamingHttpClientFilterFactory factory) {
         httpClientBuilder.appendClientFilter(predicate, factory);
     }
+
+    private TimeoutFromRequest grpcTimeout() {
+
+        return new TimeoutFromRequest() {
+            /**
+             * Return the timeout duration extracted from the GRPC timeout HTTP header if present, from the async
+             * context if present or the default timeout.
+             *
+             * @param request The HTTP request to be used as source of the GRPC timeout header
+             * @return The non-negative timeout duration which may be null
+             */
+            @Nullable
+            public Duration apply(HttpRequestMetaData request) {
+                Duration duration = readTimeoutHeader(request);
+
+                if (null != duration) {
+                    // use timeout from request
+                    return duration;
+                }
+
+                Instant deadline = AsyncContext.get(GRPC_DEADLINE_KEY);
```

Comment:
Looks like you try to extract the deadline from `AsyncContext` twice: here and in `DefaultGrpcClientCallFactory#timeoutForRequest`. Can we do it only once?

Reply:
I think it can be entirely eliminated from GrpcClientBuilder. The server operation already has a timeout filter running so client requests shouldn't need to redundantly add their own timeout filters. If the server operation times out and is  cancelled any pending client requests should be cancelled as well without them needing to have their own independent timers.